---
title: "`r params$report_title`"
output:
  html_document:
    toc: true
    toc_float: true
    theme: united
    code_folding: hide
params:
  report_title: "B-cell repertoire analysis template"
  report_author: "Anonymus"
  tagged_repertoire: "../test_data/merged_repertoire.csv"
  sample_metadata: "../test_data/samples_meta.csv"
  condition_column: "Label"
  condition_order: null
  species: "musmus"
---

## Setup

### Initialize tools

```{r message=FALSE, warning=FALSE}
# Import packages

library(tidyr)
library(purrr)
library(ggplot2)
library(ggsci)
library(ggridges)
library(plotly)
library(pheatmap)
library(DT)
library(immunarch)
```

```{r message=FALSE, warning=FALSE}
# Define function to show a nice table with download option

ALLOW_INTERACTION <- knitr::is_html_output()

show_tab <- function(tab, interaction=ALLOW_INTERACTION){
  if (interaction) {
    datatable(
      tab, extensions = "Buttons",
      options = list(
        scrollX="400px",
        dom = "Bfrtip",
        buttons = list(
          list(
            extend = "collection",
            buttons = c("csv", "excel"),
            text = "Download"
          )
        )
      )
    )
  } else {
    tab
  }
}

show_fig <- function(p, interaction=ALLOW_INTERACTION) {
  if (interaction) {
    ggplotly(p)
  } else {
    p
  }
}

color_names <- pal_npg()(9)
```

### Parse and preprocess data

```{r message=FALSE, warning=FALSE}
# Rearrange repertoire table to match Immunarch expectations

repertoire_folder <- "reconstructed_bcrs"
if (!dir.exists(repertoire_folder)) dir.create(repertoire_folder)

params$tagged_repertoire %>%
  read.csv(stringsAsFactors = FALSE) %>%
  group_split(Sample_code) %>%
  walk(function(x) {
    fn <- x$Sample_code %>%
      unique() %>%
      file.path(repertoire_folder, .)
    write.table(
      select(x, -Sample_code), paste(fn, "txt", sep="."),
      sep="\t", row.names=FALSE, quote = FALSE
    )
  })
```

```{r}
# Parse files with Immunarch

immdata <- repLoad(repertoire_folder)
unlink(repertoire_folder, recursive=TRUE, force=TRUE)
```

```{r}
# Add metadata

raw_metadata <- read.csv(params$sample_metadata)
raw_metadata$Condition <- raw_metadata[[params$condition_column]]
if(!is.null(params$condition_order)) {
  condition_order <- raw_metadata$Condition  %>%
    params$condition_order %>%
    strsplit(",") %>%
    unlist()
  raw_metadata$Condition <- factor(raw_metadata$Condition, levels=condition_order)
}
immdata$meta <- left_join(immdata$meta, raw_metadata, by=c(Sample="SampleID"))
N_CONDITION <- length(unique(raw_metadata$Condition))
```

```{r message=FALSE, warning=FALSE}
# Create long table for tidy operations

clonotype_freq_tab <- immdata$data %>%
  imap(~mutate(.x, Sample = .y)) %>%
  bind_rows() %>%
  left_join(immdata$meta, by="Sample") %>%
  mutate(
    logFreq = log(Proportion),
    receptorChain = substr(V.name, 1, 3)
  )
```

## Clonotype frequencies

### Clones recovered

```{r message=FALSE, warning=FALSE}
exp_vol <- repExplore(immdata$data, .method = "volume")
vis(exp_vol, .by = c("Condition"), .meta = immdata$meta) +
  scale_fill_manual(values=rev(color_names[1:N_CONDITION]))
```

```{r message=FALSE, warning=FALSE}
clonotype_summary <- clonotype_freq_tab %>%
  group_by(Sample, Condition, receptorChain) %>%
  summarise(
    Proportion_in_Sample = sum(Proportion),
    Mean_estimated_size  = mean(Proportion),
    Number_of_Clonotypes = n()
  )

show_tab(clonotype_summary)
```

```{r message=FALSE, warning=FALSE}
p <- clonotype_summary %>%
  ggplot(aes(x=Sample, y=Proportion_in_Sample, fill=receptorChain)) + 
    geom_bar(position="stack", stat="identity") +
    scale_fill_npg() +
    theme_bw() +
    theme(
      axis.text.x = element_text(angle=30, vjust=0.5, hjust=1)
    ) +
    labs(title="Proportion of Ig chains in samples", x="", y="Proportion")

show_fig(p)
```

```{r message=FALSE, warning=FALSE}
p <- clonotype_summary %>%
  ggplot(aes(x=Sample, y=Number_of_Clonotypes, fill=receptorChain)) + 
    geom_bar(position="stack", stat="identity") +
    scale_fill_npg() +
    theme_bw() +
    theme(
      axis.text.x = element_text(angle=30, vjust=0.5, hjust=1)
    ) +
    labs(title="Number of individual clonotypes", x="", y="Clonotype")

show_fig(p)
```

### Distribution of proportions

```{r message=FALSE, warning=FALSE}
exp_cnt <- repExplore(immdata$data, .method = "count")
vis(exp_cnt) +
  scale_color_npg()
```

```{r message=FALSE, warning=FALSE}
display_clonotype_proportion_distribution <- function(in_tab, plot_title, chain_name=NULL, n_conditions=N_CONDITION) {
  p <- in_tab %>%
    filter(if(is.null(chain_name)) TRUE else receptorChain == !!chain_name) %>%
    ggplot(aes(x=Proportion)) +
      geom_density(aes(color=Condition)) +
      scale_x_log10() +
      scale_color_manual(values=rev(color_names[1:n_conditions])) +
      theme_bw() +
      labs(title=plot_title, x="Proportion", y="Density")
  
  show_fig(p)
}

display_clonotype_proportion_distribution(
  clonotype_freq_tab, "Distribution of clonotype proportions"
)
```

```{r message=FALSE, warning=FALSE}
display_clonotype_proportion_distribution(
  clonotype_freq_tab, "Distribution of IgH clonotype proportions", "IGH"
)
```

```{r message=FALSE, warning=FALSE}
display_clonotype_proportion_distribution(
  clonotype_freq_tab, "Distribution of IgK clonotype proportions", "IGK"
)
```

```{r message=FALSE, warning=FALSE}
display_clonotype_proportion_distribution(
  clonotype_freq_tab, "Distribution of IgL clonotype proportions", "IGL"
)
```

### Most abundant clones

```{r message=FALSE, warning=FALSE}
find_top_abundant_clones <- function(in_tab, chain_name=NULL) {
  in_tab %>%
    filter(if(is.null(chain_name)) TRUE else receptorChain == !!chain_name) %>%
    group_by(Condition, CDR3.nt) %>%
    mutate(Proportion = mean(Proportion)) %>%
    ungroup() %>%
    arrange(desc(Proportion)) %>%
    group_by(Condition) %>%
    group_split() %>%
    map(head, 30) %>%
    bind_rows() %>%
    distinct(CDR3.aa, CDR3.nt)
}

show_top_clone_proportions <- function(in_tab, top_clones, plot_title, chain_name=NULL) {
  in_tab %>%
    filter(if(is.null(chain_name)) TRUE else receptorChain == !!chain_name) %>%
    filter(CDR3.nt %in% top_clones$CDR3.nt) %>%
    select(Proportion, CDR3.aa, Condition) %>%
    mutate(
      CDR3.aa = factor(CDR3.aa, levels=unique(top_clones$CDR3.aa))
    ) %>%
    plot_ly(
      x = ~CDR3.aa, y = ~Proportion, color = ~Condition, type = "box",
      boxpoints = FALSE, width=1200, height=800,
      colors=rev(color_names[1:N_CONDITION])
    ) %>%
    layout(
      xaxis = list(title="", tickangle = 30, tickfont=list(size=12)),
      yaxis = list(title="Proportion", tickfont=list(size=12)),
      boxmode = "group",
      legend = list(orientation="h", xanchor="center", x=0.5),
      title = plot_title
    )
}

display_top_abundant_clonotypes <- function(in_tab, title, chain_name=NULL) {
  top_clones <- find_top_abundant_clones(in_tab, chain_name)
  show_top_clone_proportions(in_tab, top_clones, title, chain_name)
}

display_top_abundant_clonotypes(clonotype_freq_tab, "Top overall clonotypes")
```

```{r message=FALSE, warning=FALSE}
display_top_abundant_clonotypes(clonotype_freq_tab, "Top IgH clonotypes", "IGH")
```

```{r message=FALSE, warning=FALSE}
display_top_abundant_clonotypes(clonotype_freq_tab, "Top IgK clonotypes", "IGK")
```

```{r message=FALSE, warning=FALSE}
display_top_abundant_clonotypes(clonotype_freq_tab, "Top IgL clonotypes", "IGL")
```

## Differential abundance of clonotypes

### Top clones

```{r message=FALSE, warning=FALSE}
wrap_wilcox_stats <- function(data, condition_1, condition_2, condition_col="Condition", valuecol="Proportion") {
  skip_to_next <- FALSE
  data[[condition_col]] = as.character(data[[condition_col]] )
  x_vec = unlist(data[data[[condition_col]] == condition_1, valuecol])
  y_vec = unlist(data[data[[condition_col]] == condition_2, valuecol])
  tryCatch(wilcox.test(x_vec, y_vec), error = function(e) { skip_to_next <<- TRUE })
  if(skip_to_next) {
    stat <- list(p.value=1)
  } else {
    stat <- wilcox.test(x_vec, y_vec)
  }
    data.frame(
      Comparison = paste(condition_2, "vs.", condition_1, sep=" "),
      CDR3.aa = unique(data$CDR3.aa),
      Median_1 = median(y_vec, na.rm=TRUE),
      Median_2 = median(x_vec, na.rm=TRUE),
      FC = median(y_vec, na.rm=TRUE) / median(x_vec, na.rm=TRUE),
      p.value = stat$p.value,
      stringsAsFactors = FALSE
    )
}

find_top_differential_clonotypes <- function(in_tab) {
  
  compared_conditions <- raw_metadata$Condition %>%
    unique() %>%
    gtools::combinations(N_CONDITION, 2, .) %>%
    data.frame() %>%
    setNames(c("condition_1", "condition_2"))
  
  proportion_tab <- in_tab %>%
    group_by(Sample, CDR3.aa) %>%
    mutate(Proportion = sum(Proportion)) %>%
    distinct(Sample, Condition, CDR3.aa, Proportion) %>%
    group_by(CDR3.aa) %>%
    group_split() %>%
    map(function(x) {
      compared_conditions %>%
        pmap(function(y, condition_1, condition_2){
          wrap_wilcox_stats(x, condition_1, condition_2)
        }) %>%
        bind_rows()
    }) %>%
    bind_rows()
  
  proportion_tab
    
}

differential_clono_tab <- find_top_differential_clonotypes(clonotype_freq_tab)

differential_clono_tab %>%
  filter(!is.na(FC)) %>%
  arrange(p.value) %>%
  show_tab()
```

### Overview of statistics

```{r message=FALSE, warning=FALSE}
p <- differential_clono_tab %>%
  filter(!is.na(FC)) %>%
  mutate(
    Sign = ifelse(p.value < 0.05 & (FC > 2 | FC < 0.5), "Promising", "Noise"),
    Sign = factor(Sign, levels=c("Promising", "Noise"))
  ) %>%
  ggplot(aes(x=log2(FC), y=-log10(p.value), color=Sign, label=CDR3.aa)) +
  geom_point() +
  facet_grid(Comparison ~ .) +
  scale_color_npg() +
  theme_bw() +
  labs(
    title="Volcano plot of clonotype differences",
    x="log2(FoldChange)", y="-log10(p)", color=""
  )

show_fig(p)
```

### Most significant differences 

```{r message=FALSE, warning=FALSE}
top_clones <- differential_clono_tab %>%
  filter(!is.na(FC)) %>%
  group_by(Comparison) %>%
  group_split() %>%
  map(~arrange(.x, p.value, FC)) %>%
  map(head, 15) %>%
  bind_rows() %>%
  left_join(select(clonotype_freq_tab, CDR3.nt, CDR3.aa), by="CDR3.aa") %>%
  distinct(CDR3.nt, CDR3.aa)

show_top_clone_proportions(
  clonotype_freq_tab, top_clones,
  "Clonotypes differentailly present in conditions"
)
```

## Repertoire overlap

### Shared public CDR3

```{r message=FALSE, warning=FALSE}
imm_overlap_pb <- repOverlap(immdata$data, .method = "public", .verbose = F)
vis(imm_overlap_pb) +
  ggtitle("Overlap of repertoire based on publicity index")
```

### Jaccard's index

```{r message=FALSE, warning=FALSE}
imm_overlap_jc <- repOverlap(immdata$data, .method = "jaccard", .verbose = F)
vis(imm_overlap_jc) +
  ggtitle("Overlap of repertoire based on Jaccard's index")
```

```{r message=FALSE, warning=FALSE}
p <- imm_overlap_jc %>%
  repOverlapAnalysis("mds") %>%
  .$points %>%
  data.frame() %>%
  tibble::rownames_to_column("Sample") %>%
  left_join(immdata$meta, by="Sample") %>%
  ggplot(aes(x=X1, y=X2, color=Condition, label=Sample)) +
    geom_point() +
    scale_color_manual(values=rev(color_names[1:N_CONDITION])) +
    theme_bw() +
    labs(
      title="Multidimensional scaling of Jaccard's scores",
      x="Dimension 1", y="Dimension 2", color=""
    )

show_fig(p)
```

```{r message=FALSE, warning=FALSE}
repOverlapAnalysis(imm_overlap_jc, "mds+kmeans") %>% vis()
```

### Morisita's similarity index

```{r message=FALSE, warning=FALSE}
imm_overlap_mrs <- repOverlap(immdata$data, .method = "morisita", .verbose = F)
vis(imm_overlap_mrs) +
  ggtitle("Overlap of reprtoire based on the Morisita index")
```

```{r message=FALSE, warning=FALSE}
p <- imm_overlap_mrs %>%
  repOverlapAnalysis("mds") %>%
  .$points %>%
  data.frame() %>%
  tibble::rownames_to_column("Sample") %>%
  left_join(immdata$meta, by="Sample") %>%
  ggplot(aes(x=X1, y=X2, color=Condition, label=Sample)) +
    geom_point() +
    scale_color_manual(values=rev(color_names[1:N_CONDITION])) +
    theme_bw() +
    labs(
      title="Multidimensional scaling of Morisita scores",
      x="Dimension 1", y="Dimension 2", color=""
    )

show_fig(p)
```

```{r message=FALSE, warning=FALSE}
repOverlapAnalysis(imm_overlap_mrs, "mds+kmeans") %>% vis()
```

## Repertoire diversities

### Alpha diversity

```{r message=FALSE, warning=FALSE}
display_rep_diversity <- function(intab, method, column, title, n_conditions=N_CONDITION) {
  
  dt_tab <- repDiversity(intab, method, column)
  
  if (method == "chao1") {
    dt_tab <- dt_tab %>%
      data.frame() %>%
      tibble::rownames_to_column("Sample") %>%
      rename(Value=Estimator)
  }
  
  p <- dt_tab %>%
    left_join(immdata$meta, by="Sample") %>%
    ggplot(aes(x=Condition, y=Value, color=Condition, label=Sample)) +
      geom_boxplot(outlier.shape=NA) +
      geom_point(size=2, alpha=0.8, position = position_jitterdodge()) +
      scale_color_manual(values=rev(color_names[1:n_conditions])) +
      theme_bw() +
      labs(title=title, x="", y="Chao1 index", color="")
  
  show_fig(p)
}

display_rep_diversity(
  immdata$data, "div", "aa", "Alpha diversity of CDR3 amino acids"
)
```

```{r message=FALSE, warning=FALSE}
display_rep_diversity(
  immdata$data, "div", "nt", "Alpha diversity of CDR3 nucleotides"
)
```

```{r message=FALSE, warning=FALSE}
display_rep_diversity(
  immdata$data, "div", "v", "Alpha diversity of CDR3 V genes"
)
```

### Gini-Simpson

```{r message=FALSE, warning=FALSE}
display_rep_diversity(
  immdata$data, "gini.simp", "aa", "Gini-Simpson CDR3 amino acid diversity"
)
```

```{r message=FALSE, warning=FALSE}
display_rep_diversity(
  immdata$data, "gini.simp", "nt", "Gini-Simpson CDR3 nucleotide diversity"
)
```

```{r message=FALSE, warning=FALSE}
display_rep_diversity(
  immdata$data, "gini.simp", "v", "Gini-Simpson CDR3 V gene diversity"
)
```

### Cha1 scores

```{r message=FALSE, warning=FALSE}
display_rep_diversity(
  immdata$data, "chao1", "aa", "Chao1 index of CDR3 amino acid diversity"
)
```

```{r message=FALSE, warning=FALSE}
display_rep_diversity(
  immdata$data, "chao1", "nt", "Chao1 index of CDR3 nucleotide diversity"
)
```

```{r message=FALSE, warning=FALSE}
display_rep_diversity(
  immdata$data, "chao1", "v", "Chao1 index of CDR3 V gene diversity"
)
```

## Spectratypes

### Overall spectratype

```{r message=FALSE, warning=FALSE}
display_spectratype <- function(in_tab, plot_title, chain_name=NULL, n_conditions=N_CONDITION) {
  in_tab %>%
    filter(if(is.null(chain_name)) TRUE else receptorChain == !!chain_name) %>%
    mutate(
      CDR3.len = nchar(CDR3.aa),
      Sample = factor(Sample),
      Sample = factor(Sample, levels=rev(levels(Sample)))
    ) %>%
    ggplot(aes(x=CDR3.len, y=Sample, fill=Condition)) +
      geom_density_ridges(scale=1.1, alpha=0.7) +
      scale_x_log10() +
      scale_fill_manual(values=rev(color_names[1:n_conditions])) +
      theme_bw() +
      labs(title=plot_title, x="Number of CDR3 amino acids", y="")
}

display_spectratype(
  clonotype_freq_tab, "Overall spectratype of samples"
)
```

### Heavy chain

```{r message=FALSE, warning=FALSE}
display_spectratype(
  clonotype_freq_tab, "IgH spectratype of samples", "IGH"
)
```

### Light chains

```{r message=FALSE, warning=FALSE}
display_spectratype(
  clonotype_freq_tab, "IgK spectratype of samples", "IGK"
)
```

```{r message=FALSE, warning=FALSE}
display_spectratype(
  clonotype_freq_tab, "IgL spectratype of samples", "IGL"
)
```

### V gene spectratypes

```{r message=FALSE, warning=FALSE}
for(i in seq(1, length(immdata$data))){
  
  smpl <- names(immdata$data)[[i]]
  
  p <- immdata$data[[i]] %>%
  spectratype(.quant = "count", .col = "aa+v") %>%
  vis() +
    ggtitle(paste("Spectratype of ", smpl, sep=" "))
  
  print(p)
  
}
```

## Gene V usage

### Heavy chain

```{r message=FALSE, warning=FALSE}
display_top_vgenes <- function(in_tab, plot_title, chain_name, species=params$species, n_conditions=N_CONDITION) {
  
  vgene_usage_tab <- in_tab$data %>%
    geneUsage(paste(species, chain_name, sep="."), .ambig = "maj", .norm=TRUE) %>%
    pivot_longer(-Names) %>%
    rename(
      Gene = Names,
      Sample = name
    ) %>%
    left_join(in_tab$meta, by="Sample") %>%
    filter(grepl(toupper(chain_name), Gene)) %>%
    group_by(Condition) %>%
    mutate(
      meanProp = mean(value, na.rm=TRUE)
    ) %>%
    ungroup() %>%
    arrange(desc(meanProp))
  
  top_vgene <- vgene_usage_tab %>%
    distinct(Gene) %>%
    head(25) %>%
    .$Gene
  
  vgene_usage_tab %>%
    filter(Gene %in% top_vgene) %>%
    mutate(
      Gene = factor(Gene, levels=top_vgene)
    ) %>%
    plot_ly(
      x = ~Gene, y = ~value, color = ~Condition, type = "box",
      boxpoints = FALSE, width=1200, height=800,
      colors=rev(color_names[1:n_conditions])
    ) %>%
    layout(
      xaxis = list(title="", tickangle = 30, tickfont=list(size=12)),
      yaxis = list(title="Usage score", tickfont=list(size=12)),
      boxmode = "group",
      legend = list(orientation="h", xanchor="center", x=0.5),
      title = plot_title
    )
}

display_top_vgenes(immdata, "Heavy chain V genes", "ighv")
```

### Light chains

```{r message=FALSE, warning=FALSE}
display_top_vgenes(immdata, "Kappa chain V genes", "igkv")
```

```{r message=FALSE, warning=FALSE}
display_top_vgenes(immdata, "Lambda chain V genes", "iglv")
```

### V gene clusters

```{r message=FALSE, warning=FALSE}
overall_gene_usage_tab <- geneUsage(immdata$data, paste(params$species, "ighv", sep="."), .norm=TRUE)

overall_gene_usage_tab %>%
  geneUsageAnalysis(.method = "cor", .verbose = F) %>%
  vis(.title = "Gene usage correlation", .leg.title = "Cor", .text.size = 1.5)
```

```{r message=FALSE, warning=FALSE}
overall_gene_usage_tab %>%
  geneUsageAnalysis(.method = "js", .verbose = F) %>%
  vis(.title = "Jensen-Shannon Divergence of gene usage", .leg.title = "JS", .text.size = 1.5)
```

```{r message=FALSE, warning=FALSE}
overall_gene_usage_tab %>%
  geneUsageAnalysis(.method = "js+pca", .verbose = F) %>%
  vis(.by="Condition", .meta = immdata$meta) +
  scale_fill_manual(values=rev(color_names[1:N_CONDITION])) +
  scale_color_manual(values=rev(color_names[1:N_CONDITION])) +
  ggtitle("Principal Component Analysis of JS scores")
```

```{r message=FALSE, warning=FALSE}
overall_gene_usage_tab %>%
  geneUsageAnalysis(.method = "js+pca+kmeans", .verbose = F) %>%
  vis(.by="Condition", .meta = immdata$meta, .plot = "clust") +
  scale_fill_manual(values=rev(color_names[1:N_CONDITION])) +
  scale_color_manual(values=rev(color_names[1:N_CONDITION])) +
  ggtitle("K-means clustering after PCA of JS scores")
```

```{r message=FALSE, warning=FALSE}
overall_gene_usage_tab %>%
  geneUsageAnalysis("js+hclust", .verbose = F) %>%
  vis(.plot = "clust") +
    scale_color_manual(values=rev(color_names[1:N_CONDITION])) +
    ggtitle("Hierarchical clustering of samples based on JS scores")
```

## Convergence

### Overview

```{r message=FALSE, warning=FALSE}
p <- clonotype_freq_tab %>%
  group_by(Condition, Sample, CDR3.aa) %>%
  mutate(numConverged = n()) %>%
  ungroup() %>%
  ggplot(aes(x=numConverged, fill=Condition)) +
    geom_histogram(position=position_dodge()) +
    scale_fill_manual(values=rev(color_names[1:N_CONDITION])) +
    theme_bw() +
    labs(
      title="Convergence of CDR3 regions as the number of nucleotide variants",
      x="Number of nucleotide sequence variants for CDR3", y="Number of CDR3 (not normalized)")

show_fig(p)
```

### Converged proportions

```{r message=FALSE, warning=FALSE}
p <- clonotype_freq_tab %>%
  group_by(Condition, Sample, CDR3.aa) %>%
  mutate(numConverged = n()) %>%
  ungroup() %>%
  mutate(Converged = ifelse(numConverged > 1, TRUE, FALSE)) %>%
  group_by(Sample) %>%
  mutate(Convergence = sum(Converged) / n()) %>%
  ungroup() %>%
  ggplot(aes(x=Condition, y=Convergence, color=Condition, label=Sample)) +
    geom_boxplot(outlier.shape=NA) +
    geom_point(position=position_dodge()) +
    scale_y_continuous(labels = scales::percent_format(accuracy = 1)) +
    scale_color_manual(values=rev(color_names[1:N_CONDITION])) +
    theme_bw() +
    labs(
      title="Convergence as percent of CDR3 regions in sample having more than 1 CDR3.nt variants",
      x="", y="Percent of CDR3 convergent")

show_fig(p)
```

```{r message=FALSE, warning=FALSE}
p <- clonotype_freq_tab %>%
  group_by(Condition, Sample, CDR3.aa) %>%
  mutate(numConverged = n()) %>%
  ungroup() %>%
  mutate(Converged = ifelse(numConverged > 5, TRUE, FALSE)) %>%
  group_by(Sample) %>%
  mutate(Convergence = sum(Converged) / n()) %>%
  ungroup() %>%
  ggplot(aes(x=Condition, y=Convergence, color=Condition, label=Sample)) +
    geom_boxplot(outlier.shape=NA) +
    geom_point(position=position_dodge()) +
    scale_y_continuous(labels = scales::percent_format(accuracy = 1)) +
    scale_color_manual(values=rev(color_names[1:N_CONDITION])) +
    theme_bw() +
    labs(
      title="Convergence as percent of CDR3 regions in sample having more than 5 CDR3.nt variants",
      x="", y="Percent of CDR3 convergent")

show_fig(p)
```

### Convergent clone sizes

```{r}
p <- clonotype_freq_tab %>%
  group_by(Condition, Sample, CDR3.aa) %>%
  mutate(numConverged = n()) %>%
  ungroup() %>%
  ggplot(aes(x=numConverged, y=Proportion, color=Condition)) +
    geom_density_2d() +
    scale_fill_manual(values=rev(color_names[1:N_CONDITION])) +
    theme_bw() +
    labs(
      title="Clonotype sizes as a function of convergence (density contour)",
      x="Number of nucleotide sequence variants for CDR3", y="Clone size (proportion is sample)")

show_fig(p)
```

```{r}
clonotype_freq_tab %>%
  group_by(Condition, Sample, CDR3.aa) %>%
  mutate(numConverged = n()) %>%
  ungroup() %>%
  mutate(
    numConverged = case_when(
      numConverged < 2 ~ "Sinlget",
      numConverged == 2 ~ "Doublet",
      numConverged == 3 ~ "Triplet",
      numConverged < 8 ~ "Small expansion",
      numConverged < 16 ~ "Expanded",
      TRUE ~ "Strong expansion"
    ),
    numConverged = factor(
      numConverged, levels = c(
        "Sinlget", "Doublet", "Triplet", "Small expansion",
        "Expanded", "Strong expansion"
      )
    )
  ) %>%
  ggplot(aes(x=Proportion, y=numConverged, fill=Condition)) +
    geom_density_ridges(scale=1.1, alpha=0.6) +
    scale_fill_manual(values=rev(color_names[1:N_CONDITION])) +
    scale_x_log10() +
    theme_bw() +
    labs(
      title="Clonotype sizes as a function of convergence",
      x="Clone size (proportion is sample)", y="Number of nucleotide sequence variants for CDR3")
```
